File: Recorder/src/main/java/com/github/ezauton/recorder/Recording.java
package com.github.ezauton.recorder;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.github.ezauton.core.localization.Updatable;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class Recording implements SubRecording, Updatable {
    private static int recordingCounter = 0;
    @JsonProperty("recordingData")
    private Map<String, SubRecording> recordingMap = new HashMap<>();
    @JsonProperty("name")
    private String name;

    public Recording() {
        name = "Recording_" + recordingCounter++;
    }

    public Recording(SubRecording... recordings) {
        this("Recording_" + recordingCounter++, recordings);
    }

    public Recording(String name, SubRecording... recordings) {
        this.name = name;
        for (SubRecording recording : recordings) {
            recordingMap.put(recording.getName(), recording);
        }
    }

    public Recording addSubRecording(SubRecording subRecording) {
        Optional.ofNullable(subRecording).ifPresent(r -> recordingMap.put(r.getName(), r));
        return this;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String toJson() {
        return JsonUtils.toStringUnchecked(this);
    }

    public Map<String, SubRecording> getRecordingMap() {
        return recordingMap;
    }

    @Override
    public boolean update() {
        boolean ret = false;
        for (SubRecording recording : recordingMap.values()) {
            if (recording instanceof Updatable) {
                // If the update method returns true
                if (((Updatable) recording).update()) {
                    ret = true;
                }
            }
        }
        return ret;
    }

    public void save(String name) throws IOException {
        String homeDir = System.getProperty("user.home");
        java.nio.file.Path filePath = Paths.get(homeDir, ".ezauton", name);
        save(filePath);
    }

    public void save(Path filePath) throws IOException {
        Files.createDirectories(filePath.getParent());

        BufferedWriter writer = Files.newBufferedWriter(filePath);
        String json = this.toJson();

        writer.write(json);

        writer.close();

        JsonUtils.toObject(Recording.class, json);
    }
}

File: Recorder/src/main/java/com/github/ezauton/recorder/Recording.kt
package com.github.ezauton.recorder

import com.github.ezauton.core.action.SendAction
import com.github.ezauton.core.action.sendAction
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.flow.first
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

fun generateRecordingMap(recordings: List<SubRecording>): HashMap<String, SubRecording> {
  val map = HashMap<String, SubRecording>()
  for (recording in recordings) {
    map[recording.name] = recording
  }
  return map
}
@Serializable
class Recording(override val name: String, val recordingMap: Map<String, SubRecording>): SubRecording  {

  @Throws(IOException::class)
  fun save(name: String?) {
    val homeDir = System.getProperty("user.home")
    val filePath = Paths.get(homeDir, ".ezauton", name)
    save(filePath)
    println("saved file!")
  }

  @Throws(IOException::class)
  fun save(filePath: Path) {
    Files.createDirectories(filePath.parent)
    println("path $filePath")
    val writer = Files.newBufferedWriter(filePath)

    val json = format.encodeToString(this)
    writer.write(json)
    writer.close()
  }

  companion object {
    fun from(name: String, recording: List<SubRecording>): Recording {
      return Recording(name, generateRecordingMap(recording))
    }
  }
}

fun groupRecordings(name: String, vararg actions: SendAction<SubRecording>) = sendAction {
  val subRecordings = actions.map { action ->
    async {
      action.first()
    }
  }.awaitAll()

  val recording = Recording.from(name, subRecordings);

  emit(recording)
}
